lib/matplotlib/artist.py

```diff
@@ -1417,6 +1419,80 @@ def set_mouseover(self, mouseover):
     mouseover = property(get_mouseover, set_mouseover)  # backcompat.
 
 
+class ColorizingArtist(Artist):
+    def __init__(self, norm=None, cmap=None):
+        """
+        Parameters
+        ----------
+        norm : `colors.Normalize` (or subclass thereof) or str or `cm.Colorizer` or None
+            The normalizing object which scales data, typically into the
+            interval ``[0, 1]``.
+            If a `str`, a `colors.Normalize` subclass is dynamically generated based
+            on the scale with the corresponding name.
+            If `cm.Colorizer`, the norm an colormap on the `cm.Colorizer` will be used
```

Comment:
I would not overload the signature. A typical strategy would be to have the canoncial signature and create a factory classmethod for the other signature:
```
def __init__(colorizer):
    self.colorizer = colorizer

@classmethod
def from_norm_and_cmap(cls, norm, cmap):
    return cls(Colorizer(norm, cmap)
```
AFAICT, the situation here is a bit different, because users do not instantiate `ColorizingArtist` directly, and hence, they won't use the factory method. We could do one of the two options:

- only implement `def __init__(colorizer)`. If subclasses need to initialize from norm + cmap, they'd simply do
  ```
  ColorizingArtist.__init__(Colorizer(norm, cmap))
  ```
- create an alternative initializer
  ```
  def _init_from_norm_and_cmap(norm, cmap):
      self.colorizer = Colorizer(norm, cmap)
```
given that we'll only have a handfull of subclasses and the solution is quite simple, I'm leaning towards the first option.

Reply:
The first alternative makes sense to me.

However, we still need the logic where the user can pass a `Colorizer` object as part of the input to the plotting function.

I implemented this with the following function:
```py
def _get_colorizer(cmap, norm):
    """
    Passes or creates a Colorizer object.

    Allows users to pass a Colorizer as the norm keyword
    where a artist.ColorizingArtist is used as the artist.
    If a Colorizer object is not passed, a Colorizer is created.
    """
    if isinstance(norm, Colorizer):
        if cmap:
            raise ValueError("Providing a `cm.Colorizer` as the norm while "
                             "at the same time providing a `cmap` is not supported.")
        return norm
    return Colorizer(cmap, norm)
```
Such that the initiation of a `ColorizingArtist` in e.g. `Collection` is:
```py
        artist.ColorizingArtist.__init__(self, colorizer._get_colorizer(cmap, norm))
```

Reply:
Do we really want to pass a `Colorizer` via the `norm` parameter in public API? This feels very hacky. I'm inclined to add a `colorizer` parameter instead, even if that means we have mutually exclusive parameters. (Note that even with the hack, we'd have to live with partially mutually exclusive parameters, because you cannot pass `norm=mycolorizer, cmap='jet'`).

Reply:
I have no strong opinions, and am inclined to agree with you @timhoffm.
Since this is a change to the call signature of the plotting functions, we should bring it up at a weekly meeting. @story645 could you add it on the agenda for the next meeting?

(My aversion to new keyword arguments is a result of a discussion at a weekly meeting a long time ago about multivariate color mapping, and how we do not want *that* to increase the number of keyword arguments, but this is a different case entirely.)